### PR TITLE
Apply PHP 7.1 and PHP 7.2 features

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: Classes/Api/CrawlerApi.php [hunk 1]
```diff
@@ -502,4 +404,95 @@ public function getPerformanceData($start, $end, $resolution)
...
+    {
+        $crawlerProcInstructions = [];
+        if (! empty($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['crawler']['procInstructions'])) {
+            foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['crawler']['procInstructions'] as $configuration) {
+                $crawlerProcInstructions[$configuration['key']] = $configuration['value'];
+            }
+        }
+
+        return $crawlerProcInstructions;
+    }
```
Comment (TomasVotruba) [inline Classes/Api/CrawlerApi.php]: These method order changes are due to sort methods by `public`, `protected` and `private` visibility

**Comment by TomasVotruba** [comment]
Ready to review + merge :)

**Comment by tomasnorre** [comment]
Could you please have a look at the failing tests

File: Classes/CrawlStrategy/SubProcessExecutionStrategy.php [hunk 2]
```diff
@@ -39,7 +40,7 @@ class SubProcessExecutionStrategy implements LoggerAwareInterface
      */
     protected $extensionSettings;
 
-    public function __construct(ExtensionConfigurationProvider $configurationProvider = null)
+    public function __construct(?ExtensionConfigurationProvider $configurationProvider = null)
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply Classes/CrawlStrategy/SubProcessExecutionStrategy.php]: Its very nicely explained here:

https://symfonycasts.com/screencast/php7/nullable-types
"No problem, right? Refresh! ..."

**Comment by TomasVotruba** [comment]
I missed it, will check

File: Classes/CrawlStrategy/SubProcessExecutionStrategy.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomasnorre) [reply Classes/CrawlStrategy/SubProcessExecutionStrategy.php]: Thanks, Will read it later.

**Comment by tomasnorre** [comment]
Just to let you know. 

Nicole from Nimut/testing-framework told me yesterday that: 

> the only things, he cannot do right now are
> `protected function setUp(): void` and `protected function tearDown(): void`

I don't know if that it part of the problem with the tests.

**Comment by TomasVotruba** [comment]
Yes, that's also a blocker :/ the typo3 ecosystem is pretty lagging. This is 2 years old change in PHPUnit

**Comment by tomasnorre** [comment]
Not all things are going in the same speed. Most of the Core is in good shape. 
But some extensions like the crawler, needs some love. 

But as your know yourself, maintaining open source in your free time, sets limits.

**Comment by TomasVotruba** [comment]
CI fixed

**Comment by TomasVotruba** [comment]
@tomasnorre I think it's not about maintaners, but wrong split package architecture.

Let's say there are over 150 typo3 plugins. Why upgrading all of them one by one, if it's just 1 change? Now it's 150x same repeated work. Change class, PHP version what ever. Such a waste of human energy.

With Rector/framework monorepo this would be just 1 commits to them all

**Comment by tomasnorre** [comment]
I get your point, but it's think it's difficult with so many maintainers, there are more than 3000 different TYPO3 Extensions, and having this in a mono repo, were every of the contributor can push, is in my eyes no doable. It's not only a matter of keeping things up to date, but also developing new features. 

But we don't need to discuss this here. :)

**Comment by TomasVotruba** [comment]
> I get your point, but it's think it's difficult with so many maintainers, there are more than 3000 different TYPO3 Extensions, and having this in a mono repo, were every of the contributor can push, is in my eyes no doable. It's not only a matter of keeping things up to date, but also developing new features.

I don't mean having one monorepo where everyone can push. But having one person making this set that could be automated, like thsi PR :) I bet it's easier for you to just click "Merge" then for me to do the all the work, right? :)

Best of both worlds ;) control + one source knowledge

**Comment by tomasnorre** [comment]
I'm not sure I get you. How this would look like, and how the setup should be. 

You mean e.g. I have every extension cloned, and run the rector scripts over them? to have libraries etc updated on a regular basic? 

I'm quite curios about this, it could be a fine way to support the community if I could something like that up.

**Comment by TomasVotruba** [comment]
It's an idea comming from super slow upgrade of all the framework extensions.
It's faily easy to upgrade project from Symfony 4 to 5 (use any framework/version you use). But all the extensions take soo much work to upgrade. And the change is always the same.

> You mean e.g. I have every extension cloned, and run the rector scripts over them? to have libraries etc updated on a regular basic?

You don't do any work, that's the idea. Just get a PR with "Hey, new TYPO3 10 is out, here is the upgrade" and all you do is click button and accept merge request.

**Comment by tomasnorre** [comment]
> You don't do any work, that's the idea. Just get a PR with "Hey, new TYPO3 10 is out, here is the upgrade" and all you do is click button and accept merge request.

I know, but someone has to create that Pull Request, and whom should that be.. I could be that person, that's why I meant before.

**Comment by TomasVotruba** [comment]
I think it should be automated, like "who should run the tests after every commit" automated

**Comment by tomasnorre** [comment]
I think I get you know.. Will see what I can come up with to have a concept for the setup to discuss with the community.

**Pull request merged**